wikiconv/conversation_reconstruction/dataflow_main.py

```diff
@@ -15,139 +15,230 @@
+python dataflow_main.py \
+  --setup_file ./setup.py \
+  --input_state 'gs://YOUR_BUCKET/PATH_TO_INIT_STATE' \
+  --input_revisions 'gs://YOUR_BUCKET/PATH_TO_REVS*' \
+  --output_state 'gs://YOUR_BUCKET/PATH_TO_OUTPUT_STATE_TO' \
+  --output_conversations 'gs://YOUR_BUCKET/PATH_TO_OUTPUT_CONVS_TO' \
+  --runner DirectRunner
+  --project $YOUR_GCLOUD_PROJECT \
+  --job_name $NAME_OF_YOUR_JOB
+  --num_workers $NUMBER_OF_WORKERS_SUCH_AS_80
+```
+
+Note: Don't forget the quotes on the local paths, otherwise bash will interpret
+them and dataflow fails to see all the files.
+TODO(ldixon): make the input parser smarter so that it can handle this.
 
-  ```
-  cd helper_shell
-  ./reconstruct.sh
-  ```
 """
 from __future__ import absolute_import
 import argparse
 import logging
 import traceback
 import subprocess
-import resource
 import json
 import os
 from os import path
 import urllib2
-import tempfile
 import ast
-import copy
 import sys
 import time
 
 import apache_beam as beam
 from apache_beam.io import filesystems
 from apache_beam.metrics.metric import Metrics
+from apache_beam.metrics.metric import MetricsFilter
 from apache_beam.options.pipeline_options import PipelineOptions
 from apache_beam.options.pipeline_options import SetupOptions
-from construct_utils.conversation_constructor import Conversation_Constructor
+from construct_utils.reconstruct_conversation import ReconstructConversation
 
+# The max cumulative size of a page's revisions to be considered to try and
+# keep in memory when processing.
+# Measured by python string length. approx 250 MB
+CUMULATIVE_REVISION_SIZE_THERESHOLD = 250 * 1024 * 1024
 
-LOG_INTERVAL = 100
-# The max memory used in of this process KB, before warning are logged.
-MEMORY_THERESHOLD = 1000000
-REVISION_THERESHOLD = 250000000
+# Constants used
 SAVE_TO_MEMORY = 0
 SAVE_TO_STORAGE = 1
+
+# Number of times to DataFlow should retry a failed worker.
 RETRY_LIMIT = 10
 
-def get_metadata(x):
-  record_size = len(x)
-  record = json.loads(x)
+# Custom logging level so we don't have to read all the info messages, but we
+# see our printed summary statistics.
+LOG_LEVEL_OUTPUT_INFO = 25
+
+def page_indexed_metadata_of_revstring(rev_string):
+  record_size = len(rev_string)
+  record = json.loads(rev_string)
   return (record['page_id'], {'page_id': record['page_id'], 'rev_id': record['rev_id'],
                               'timestamp': record['timestamp'], 'record_size': record_size})
 
 
-def run(known_args, pipeline_args):
-  """Main entry point; defines and runs the reconstruction pipeline."""
+def index_by_page_id(s):
+  """Given a json string of a dict with a page_id key, pair the page_id with
+  the object."""
+  return (json.loads(s)["page_id"], json.loads(s))
+
+
+def index_by_rev_id(s):
+  """Given a json string of a dict with a rev_id key, pair the rev_id with
+  the string"""
+  return (json.loads(s)["rev_id"], s)
 
-  if known_args.testmode:
-    pipeline_args.append('--runner=DirectRunner')
```

Comment:
Might be good to keep an option to run as LocalRunner, as a different test choice.

Reply:
The runner argument is now provided when we run the script (which assumes more dataflow knowledge (but we can document it more in the README), we we don't need this in the code as the person running the command can choose the runner themselves.